core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -144,12 +160,39 @@ private void createActiveTransportFuture() {
     }
   }
 
+  @VisibleForTesting
+  long currentTimeMillis() {
+    return System.currentTimeMillis();
+  }
+
   // Can only be called when shutdown == false
   @GuardedBy("lock")
   private void scheduleConnection() {
     Preconditions.checkState(!shutdown, "Already shut down");
     Preconditions.checkState(reconnectTask == null || reconnectTask.isDone(),
         "previous reconnectTask is not done");
+
+    final int currentAddressIndex = nextAddressIndex;
+    List<SocketAddress> addrs = addressGroup.getAddresses();
+    final SocketAddress address = addrs.get(currentAddressIndex);
+    nextAddressIndex++;
+    if (nextAddressIndex >= addrs.size()) {
+      nextAddressIndex = 0;
+    }
+
+    // If it's the second time reaching headIndex, we have failed the entire list in a
+    // row, then nextHeadAttemptMillis will apply.
+    final long delay = (currentAddressIndex == headIndex)
```

Comment:
This backoff logic seems more complex than necessary, especially because of `currentTimeMillis()` and `nextHeadAttemptMillis`.

Why not something closer to:
```java
if (headIndex == -1) {
  headIndex = currentAddressIndex;
  // First connect attempt, or the first attempt since last successful connection.
  // UNMODIFIED
  reconnectPolicy = backoffPolicyProvider.get();
  createTransportRunnable.run();
  reconnectTask = null;
} else if (currentAddressIndex != headIndex) {
  // Another attempt, but to a different IP. Don't backoff.
  createTransportRunnable.run();
  reconnectTask = null;
} else if (currentAddressIndex == headIndex) {
  // Back to beginning of list. Trigger backoff.
  // UNMODIFIED
  long delayMillis = reconnectPolicy.nextBackoffMillis();
  reconnectTask = scehduledExecutor.schedule(
      createTransportRunnable, delayMillis, TimeUnit.MILLISECONDS);
}